test/print_test_stats.py

```diff
@@ -125,6 +131,29 @@ def send_report(reports):
     )
     r.raise_for_status()
 
+def send_report_to_s3(reports):
+    sha1 = os.environ.get('CIRCLE_SHA1')
+    now = datetime.datetime.utcnow().isoformat()
+    key = f'test_time/{sha1}/{now}Z.json'  # Z meaning UTC
+    obj = boto3.resource('s3').Object('ossci-metrics', key)
+    obj.put(Body=json.dumps({
+        'meta': build_info(),
+        'suites': {
+            name: [
+                # making this a list of dicts rather than a dict from
+                # names to times, because a) I'm not sure whether the
+                # case names are unique, and b) this way the time units
+                # are clearly labeled as "seconds" by their JSON key
```

Comment:
question: what would be the consumer of these data look like? it might help understand this question first before designing what's the best storage format.

Reply:
good point; the goal of using this data (as mentioned in the PR description above) is to add an additional step at the end of the `pytorch_linux_test` job which uses this data (for the current commit and the base revision on `master`) to decide whether there's a test time regression, and if so, modify the Dr CI comment; or is that not what you were asking about specifically?

Reply:
Yes. I guess the question here is 
(1) what is the object we aer comparing? per job basis, per test case or per test suite? these all have pro/cons. 
(2) what baseline are we comparing? is it medium/avg/p70 of master CI results, or is it comparing more higher-order data trends? 
(3) what is the threshold for alerting? 

For example, if we are looking at job level, then we want to split the storage by job..

Reply:
Good questions @walterddr. Regarding your question 1, my understanding is that we have two goals for this:

1. warn when the total time spent on testing increases by too much, since this corresponds to CI cost
2. warn when the time for the critical path (or maybe any path?) increases by too much, since this corresponds to time-to-signal

Since all the jobs are run in parallel, I'm guessing the per-job time corresponds to a "path" as I'm using the word for (2) above, but I would also guess that the "total time" for (1) would be the sum of all jobs for a commit, which is why I'm thinking maybe the job should come after the commit in the key. I'm envisioning a process something like this:

1. find the base commit `${base_sha}` on `master`
2. get all objects in the bucket whose key has the prefix `test_time/${base_sha}/`
3. for total time, add up the times from all those objects (or maybe filter by taking only the most recently-timestamped object under each job?)
4. for time-to-signal, first find the job `${current_job}` that is currently running and drill down further to get just the objects with prefix `test_time/${base_sha}/${current_job}/`

I'm not yet very experienced with S3, but I believe that steps 2 and 3 would be more difficult if we put the job name before the SHA1.

Regarding your question 3 (and part of your question 2), here's a snippet from my notes about my meeting with @malfet on 2020-11-17:

> - use a 5% or 10% threshold; don’t warn on every time increase
> - use difference divided by the minimum of (old, new)

Also for your question 2, if we want to get more stable running averages for master CI results, should we query the S3 bucket when uploading and roll those results together into the current upload, or should we just have the upload use the data from the current commit and then have the PR-commenting process look at multiple commits at once when looking for regressions? I'm kind of leaning toward the latter, since it would allow us to be more flexible in how we find regressions in the future, without repeatedly changing the format of the data we store in S3.